toolchain/lex/lex.cpp

```diff
@@ -870,6 +895,46 @@ auto Lexer::LexCommentOrSlash(llvm::StringRef source_text, ssize_t& position)
   CARBON_CHECK(result, "Failed to form a token!");
 }
 
+auto Lexer::StartDumpSemIRRange(const char* diag_loc) -> void {
+  EndDumpSemIRRangeIfIncomplete(diag_loc);
+
+  // The start here will be the next token, which may be FileEnd. The end will
+  // be assigned by either AddDumpSemIREnd or, if invalid,
+  // EndDumpSemIRRangeIfIncomplete.
+  buffer_.dump_sem_ir_ranges_.push_back(
+      {.start = TokenIndex(buffer_.size()), .end = TokenIndex::None});
+}
+
+auto Lexer::EndDumpSemIRRange(const char* diag_loc) -> void {
+  if (buffer_.dump_sem_ir_ranges_.empty() ||
+      buffer_.dump_sem_ir_ranges_.back().end != TokenIndex::None) {
```

Comment:
When does the `end != None` case come up?

Reply:
Note, concretely, this is the second error in fail_misordered.carbon